experiments/srupp_experiments/train_enwik8.py

```diff
@@ -0,0 +1,392 @@
+        loss = F.cross_entropy(output, y.view(-1), reduction='none')
+        loss = loss.view(y.size(0), y.size(1))
+        return loss, hidden, memory
+
+    def init_hidden(self, batch_size):
+        weight = next(self.parameters()).data
+        zeros = weight.new(self.depth, batch_size, self.n_d).zero_()
+        return zeros
+
+
+def read_corpus(path, num_test_symbols=5000000):
+    raw_data = open(path).read()
+    raw_data = np.fromstring(raw_data, dtype=np.uint8)
+    unique, data = np.unique(raw_data, return_inverse=True)
+    train_data = data[: -2 * num_test_symbols]
+    valid_data = data[-2 * num_test_symbols: -num_test_symbols]
+    test_data = data[-num_test_symbols:]
+    return train_data, valid_data, test_data, unique
+
+
+def create_batches(data_ids, batch_size, n_nodes=1, rank=0, device='cpu'):
+    N = len(data_ids)
+    L = (N-1) // (batch_size * n_nodes) * batch_size * n_nodes
+    x = np.copy(data_ids[:L].reshape(n_nodes, batch_size, -1)[rank].T)
+    y = np.copy(data_ids[1:L+1].reshape(n_nodes, batch_size, -1)[rank].T)
+    x, y = torch.from_numpy(x), torch.from_numpy(y)
+    x, y = x.contiguous(), y.contiguous()
+    x, y = x.to(device), y.to(device)
+    return x, y
+
+
+def calc_norm(lis):
+    l2_sum = sum(x.norm()**2 for x in lis)
+    return l2_sum**0.5
+
+
+def eval_model(model, valid):
+    with torch.no_grad():
+        model.eval()
+        args = model.args
+        batch_size = valid[0].size(1)
+        total_loss = 0.0
+        unroll_size = args.eval_unroll_size or args.unroll_size
+        model.set_mask(unroll_size, same_length=True)
+        hidden = model.init_hidden(batch_size)
+        memory = [None] * args.depth
+        N = (len(valid[0])-1)//unroll_size + 1
+        for i in range(N):
+            x = valid[0][i*unroll_size:(i+1)*unroll_size]
+            y = valid[1][i*unroll_size:(i+1)*unroll_size]
+            loss, hidden, memory = model(x, y, hidden, memory=memory)
+            total_loss += loss.sum()
+        avg_loss = total_loss.item() / valid[1].numel()
+        ppl = np.exp(avg_loss)
+        model.train()
+        model.set_mask(args.unroll_size, same_length=False)
+        return ppl, avg_loss
+
+
+def copy_model(model):
+    states = model.state_dict()
+    for k in states:
+        v = states[k]
+        states[k] = v.clone().cpu()
+    return states
+
+
+def set_seed(seed):
+    random.seed(seed)
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)
+
+
+def main(args):
+
+    if args.local_rank == 0:
+        log_path = "{}_{}".format(args.log, random.randint(1, 100))
+        train_writer = SummaryWriter(log_dir=log_path+"/train")
+        dev_writer = SummaryWriter(log_dir=log_path+"/dev")
+
+    # set up distributed training
+    torch.cuda.set_device(args.local_rank)
+    device = torch.device("cuda", args.local_rank)
+    torch.distributed.init_process_group(backend="nccl")
+    set_seed(1234)
```

Comment:
maybe seed should be a command-line option, so we can try with different seeds?

Reply:
will do.